### PR TITLE
Adds support for ava.config.js files

### LINKED ISSUE #520: Adding support for a dedicated configuration file
I'm not sure how viable this is currently due to the way pkg-conf is walking up until it finds a `package.json` file, but it would be nice to be able to set ava config in an `.avarc` file in the root of the project.

I'm not a huge fan of keeping config in my package.json or of overly verbose cli commands in `scripts` config. This is also pretty common practice with other tooling frameworks so I don't think it's too crazy of an idea.

I think an initial approach could be to check the current working directory for a `.avarc` file and if not found fall back to `pkgConf.sync`.

I would be more than willing to put together a PR if this seems like something you find appealing.

### PR DESCRIPTION
**Pull request by good-idea**
Fixes #520 

 - [x] When invoked, AVA walks the file hierarchy looking for a package.json file
 - [x] AVA checks whether a sibling ava.config.js file exists
 - [x] If the package.json file contains an "ava" configuration, and ava.config.js exists, AVA will exit with a warning about ambiguous configuration
 - [x] Otherwise the package.json configuration takes precedence over ava.config.js
 - [x] ava.config.js is loaded through esm so can be either a CJS or ESM file. The loader should support ESM transpilation by looking for a thruthy __esModule property: if set then the value of the default property is used
 - [x] An actual configuration object may be exported, or a factory function
 - [x] If a factory function is exported it'll be called with the project directory (for future compatibility). It must return a configuration object
 - [x] The configuration object must not have a then property
 - (These are the same rules as for .babelrc.js files in Babel 7)
 - [x] ~Smart JS/JSON filetype in babel-config.js~
 - [x] Update docs
 - [x] Explore [`esm` options](https://www.npmjs.com/package/esm#options) so we don't have to support the module.exports workarounds etc

### CODE REVIEW

**Comment by pvdlg** [comment]
Have you checked [cosmiconfig](https://github.com/davidtheclark/cosmiconfig)? It would fulfill a lot of the requirements, and the one missing might be implement over there.

**Comment by novemberborn** [comment]
We'd end up using very little of `cosmiconfig`. It's relatively straight-forward to hook this up directly in AVA.

@good-idea thanks for your PR. I've been quite busy lately so I make need another few days before I can get back to you.

**Comment by good-idea** [comment]
@novemberborn No hurry, I don't think I'll be able to chip away at this until the end of the week.

This is my first contribution to OSS 🤓 so any feedback is welcome!

File: lib/load-config.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+const findPackageDotJson = findFileUp('package.json');
+
+const findConfigFile = findFileUp(/ava\.config(\.\w+)?\.js/);
+
+const findConfigFiles = () => {
+	const packageFile = findPackageDotJson();
+	const configFile = findConfigFile(
+		packageFile ? path.dirname(packageFile) : process.cwd(),
```
Comment (novemberborn) [inline lib/load-config.js]: Let's simplify this. We can keep the current `pkg-conf` logic to find the `package.json`, if any. Afterwards, the `ava.config.js` file *must* be in the `projectDir`. We won't walk the file system to find it. (This means the `findFileUp` logic above is not necessary.)

File: test/cli.js [hunk 2]
```diff
@@ -40,7 +40,9 @@ function execCli(args, opts, cb) {
 
 		child.on('close', (code, signal) => {
 			if (code) {
-				const err = new Error(`test-worker exited with a non-zero exit code: ${code}`);
+				const err = new Error(
+					`test-worker exited with a non-zero exit code: ${code}`,
```
Comment (novemberborn) [inline test/cli.js]: Looks like you're using Prettier. Could you exclude unrelated changes from the PR?

**Comment by novemberborn** [comment]
@good-idea if you could merge in (or rebase on) master, that should fix some of the CI issues.

**Comment by good-idea** [comment]
@novemberborn thanks, I'll working on this later today!

File: test/cli.js (hunk 2 above)
Comment (good-idea) [reply test/cli.js]: Got it!

**Comment by good-idea** [comment]
@novemberborn I think this just about does it, let me know if you see anything else that should change!

File: lib/load-config.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+'use strict';
+const path = require('path');
+const fs = require('fs');
+// ESM seems to break `path` and `fs`, require them first before setting it up..
```
Comment (novemberborn) [inline lib/load-config.js]: `esm` shouldn't impact `require()` calls though. Happy to see the Node.js builtins loaded first, but your comment hints at an `esm` bug. Is this still happening for you?

File: lib/load-config.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+const requireESM = require('esm')(module);
+const pkgConf = require('pkg-conf');
+
+const loadConfigFile = projectDir => {
+	const loaded = requireESM(path.resolve(projectDir, 'ava.config.js'));
+	const config = (loaded.__esModule) ? loaded.default : loaded;
```
Comment (novemberborn) [inline lib/load-config.js]: Does `esm` take care of this?

File: lib/load-config.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	if (config.then) {
+		throw new Error('Config file must not return a promise');
+	}
+	return (typeof config === 'function') ? config(projectDir) : config;
```
Comment (novemberborn) [inline lib/load-config.js]: Let's pass `{projectDir}`, so we can extend this with other options later.

File: lib/load-config.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+		throw new Error('Warning: You have configuration settings in both package.json and ava.config.js. Please use only one or the other.');
+	}
+
+	const config = configFileExists ? loadConfigFile(projectDir) : packageConf;
+
+	const _meta = {
+		projectDir
+	};
+
+	return Object.assign(config, defaults, {_meta});
```
Comment (novemberborn) [inline lib/load-config.js]: This copies the `defaults` *into* `config`. The `_meta` bit seems unnecessary to me.

Better would be `Object.assign({}, defaults, config, {projectDir})`.

File: profile.js [hunk 7]
```diff
@@ -38,9 +38,9 @@ const conf = pkgConf.sync('ava', {
 	}
 });
 
-const filepath = pkgConf.filepath(conf);
-const projectDir = filepath === null ? process.cwd() : path.dirname(filepath);
+console.log(conf);
```
Comment (novemberborn) [inline profile.js]: Don't forget to remove 😉

**Comment by novemberborn** [comment]
Oh, one more thing. The `fileType` values in https://github.com/avajs/ava/blob/718349ece833c30052138862785a4e7ee7bc02e6/lib/babel-config.js#L136 need to change to `'JS'` when the config originates from a `.js` file. I'm happy to fix and test that myself though.

File: lib/load-config.js (hunk 3 above)
Comment (good-idea) [reply lib/load-config.js]: I haven't looked into why, but if I set up ESM first like so:

File: lib/load-config.js (hunk 4 above)
Comment (good-idea) [reply lib/load-config.js]: no, it doesn't seem to. There's a test for this that fails if you don't do the check for `__esModule`.

File: lib/load-config.js (hunk 6 above)
Comment (good-idea) [reply lib/load-config.js]: ✅

File: lib/load-config.js (hunk 5 above)
Comment (good-idea) [reply lib/load-config.js]: ✅

File: profile.js (hunk 7 above)
Comment (good-idea) [reply profile.js]: ✅

**Comment by good-idea** [comment]
@novemberborn This last update should address the notes in your review -- I didn't look into the fileType issue in babel-conifg yet, though. I updated the initial PR description to include a todo for this.

File: lib/load-config.js (hunk 3 above)
Comment (novemberborn) [reply lib/load-config.js]: This works for me, Node.js 8.11.0, and `esm@3.0.22`.

File: lib/load-config.js (hunk 5 above)
Comment (novemberborn) [reply lib/load-config.js]: Ah, I meant doing `config({projectDir})`, not `loadConfigFile({projectDir})`.

File: lib/load-config.js (hunk 3 above)
Comment (good-idea) [reply lib/load-config.js]: `esm@3.0.15` was installed, I updated to `3.0.22` and it now works as expected.

File: lib/load-config.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+require = require('esm')(module); // eslint-disable-line no-global-assign
```
Comment (novemberborn) [inline lib/load-config.js]: Let's do `const requireEsm` though. We don't need to change any other imports.

File: package.json [hunk 9]
```diff
@@ -92,6 +92,7 @@
 		"emittery": "^0.3.0",
 		"empower-core": "^0.6.1",
 		"equal-length": "^1.0.0",
+		"esm": "^3.0.22",
```
Comment (novemberborn) [inline package.json]: The `package-lock.json` diff seems surprisingly large given this one change. Might be good to `git checkout master package-lock.json` and `npm i` again. (At least, I think that's the right `git` incantation.)

File: lib/load-config.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+const path = require('path');
+const fs = require('fs');
+// ESM seems to break `path` and `fs`, require them first before setting it up..
```
Comment (novemberborn) [inline lib/load-config.js]: This comment can be removed now.

File: lib/load-config.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+const pkgConf = require('pkg-conf');
+
+const loadConfigFile = projectDir => {
+	const loaded = require(path.resolve(projectDir, 'ava.config.js'));
+	const config = (loaded.__esModule) ? loaded.default : loaded;
```
Comment (novemberborn) [inline lib/load-config.js]: Would be good to have a fixture with an `__esModule` export.

Let's ignore the edge case of `export default {__esModule: true, default: {}}` 😉

File: lib/load-config.js (hunk 8 above)
Comment (good-idea) [reply lib/load-config.js]: no problem, ✅

File: package.json (hunk 9 above)
Comment (good-idea) [reply package.json]: Ok, better now: from 6,951 to 266

**Comment by good-idea** [comment]
@novemberborn Just added a new commit with info in the docs. Let me know what you think!
I've also updated the checklist in the original comment.
I don't think I'll have time to dive into babel-config.js until next weekend. I'm happy to, though -- but if so could you give me a little direction on how you think it would best be implemented?

File: lib/load-config.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+const requireEsm = require('esm')(module); // eslint-disable-line no-global-assign
```
Comment (sindresorhus) [inline lib/load-config.js]: I think think `// eslint-disable-line no-global-assign` is needed.

File: readme.md [hunk 13]
```diff
@@ -287,6 +287,61 @@ Arguments passed to the CLI will always take precedence over the configuration i
...
+
+To use an `ava.config.js` file:
+
+ 1. It must be in the same directory as your `package.json`.
+ 2. Your `package.json` must not contain an `ava` property.
+
+The config file can use either ESM or CommonJS syntax. Both of these are valid:
+
+```js
+module.exports = {
```
[the remark this answers is not in the document]
Comment (novemberborn) [reply readme.md]: @sindresorhus I can't quite recall. I can see users outputting this file from a build step though, so it'd be annoying if that wouldn't work.

But we could start there. I think `esm` can be configured to not allow CJS.

File: lib/load-config.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+const requireEsm = require('esm')(module);
+const path = require('path');
+const fs = require('fs');
+const pkgConf = require('pkg-conf');
+
+const loadConfigFile = projectDir => {
+	const loaded = requireEsm(path.resolve(projectDir, 'ava.config.js'));
+	const config = (loaded.__esModule) ? loaded.default : loaded;
```
Comment (novemberborn) [inline lib/load-config.js]: I want to play around with the [`esm` options](https://www.npmjs.com/package/esm#options). I *think* we can make it always treat these files as ESM, so we don't have to support the `module.exports` workarounds etc. Would be neat to start with that. What do you think @sindresorhus?

File: lib/load-config.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	if (config.then) {
+		throw new Error('Config file must not return a promise');
+	}
+	return (typeof config === 'function') ? config({projectDir}) : config;
```
Comment (novemberborn) [inline lib/load-config.js]: The `config({projectDir})` return value *also* must not be a promise. Happy to fix this up myself.

File: readme.md [hunk 16]
```diff
@@ -287,6 +287,63 @@ Arguments passed to the CLI will always take precedence over the configuration i
...
+};
+```
+
+```js
+export default {
+	// …
+};
+```
+
+The config file may also export a factory function, which will be called with an object containing a `projectDir` property. This function must return a config object, and cannot be a promise.
```
Comment (novemberborn) [inline readme.md]: Perhaps:

> This function must return a config object, it mustn't return a promise.

File: readme.md [hunk 17]
```diff
@@ -287,6 +287,63 @@ Arguments passed to the CLI will always take precedence over the configuration i
...
+
+```js
+const config = ({projectDir}) => {
+	if (projectDir === '...') {
```
Comment (novemberborn) [inline readme.md]: `projectDir` will be an absolute path, so it'd be good if the example didn't show dots which imply it's somehow relative.

Providing the directory is useful if you have share AVA config amongst projects, e.g. in a work setting or a monorepo. You could then differentiate between projects when necessary. Perhaps this is something we should cover in a recipe, outside of this PR.

File: readme.md [hunk 18]
```diff
@@ -287,6 +287,63 @@ Arguments passed to the CLI will always take precedence over the configuration i
...
+
+	return {
+		// Config B
+	};
+}
+
+export default config;
+```
+
+Or, return different configuration based on the environment: `NODE_ENV=development ava`
```
Comment (novemberborn) [inline readme.md]: I'm not sure about advocating AVA to be run with different `NODE_ENV` settings. It'd be more interesting to detect CI environments — although at that point we should forward the detection to the function. Perhaps remove for now and track in a follow-up issue?

File: lib/load-config.js (hunk 14 above)
Comment (sindresorhus) [reply lib/load-config.js]: 👍

**Comment by good-idea** [comment]
@novemberborn I made some changes to address the above -- there another big diff with `package-lock.json`, so I did the same as before. Let me know if everything is looking OK.

**Comment by novemberborn** [comment]
@good-idea I haven't had a chance yet to look into this. I just realized we need some changes with Babel (https://github.com/avajs/ava/issues/1789) so I might try and land that before doing my work here.

Anyway this is super close so once that unblocks we should be good to go.

**Comment by novemberborn** [comment]
@good-idea just wanted to say I haven't forgotten about this 😄  #1789 is pretty much done so I'm hoping to land that next weekend. I'll then try and get this in as well before cutting a release.

**Comment by good-idea** [comment]
@novemberborn No hurry on my end! I've been using my own fork for now :)
Looking forward to the next release

**Comment by novemberborn** [comment]
@good-idea I've merged latest `master` and pushed documentation improvements and some cleanup of the config loading. We're also now using just ESM.

There's a few scenarios that are lacking tests. Once I've done that I'll land this 🎉
